File: src/com/easy/SetMismatch_645.java
/**
 * 
 */
package com.easy;

import java.util.HashSet;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class SetMismatch_645 {
    // 2020年2月13日
    // Use swap index:
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 2 ms, faster than 95.10%
    // Memory Usage: 42.8 MB, less than 14.29%
    public int[] findErrorNums(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            while (nums[i] != (i + 1) && nums[nums[i] - 1] != nums[i]) {
                swap(nums, i, nums[i] - 1);
            }
        }
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != (i + 1)) {
                return new int[] { nums[i], i + 1 };
            }
        }

        return null;
    }

    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    // 2020年2月13日
    // Use Hash Set:
    // Time Complexity: O(2n) = O(n)
    // Space Complexity:O(n)
    // Runtime: 14 ms, faster than 33.39%
    // Memory Usage: 42.9 MB, less than 14.29%
    public int[] findErrorNums2(int[] nums) {
        int[] mResult = new int[2];
        HashSet<Integer> mSet = new HashSet<>();

        for (int num : nums) {
            if (mSet.contains(num)) {
                mResult[0] = num;
            }
            mSet.add(num);
        }
        for (int i = 0; i < nums.length; i++) {
            if (mSet.contains(i + 1)) {
                mSet.remove(i + 1);
            } else {
                mResult[1] = i + 1;
                break;
            }
        }

        return mResult;
    }
}

File: Golang_src/easy/setMismatch_645.go
package easy

// Use Constant Space:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 48 ms, faster than 57.41%
// Memory Usage: 6.8 MB, less than 44.44%
func findErrorNums(nums []int) []int {
	for i := range nums {
		for nums[i] != (i+1) && nums[nums[i]-1] != nums[i] {
			nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
		}
	}
	for i := range nums {
		if nums[i] != (i + 1) {
			return []int{nums[i], i + 1}
		}
	}
	return []int{}
}

// Use XOR:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 32 ms, faster than 67.74%
// Memory Usage: 6.6 MB, less than 83.87%
func findErrorNums2(nums []int) []int {
	xor := 0
	for _, v := range nums {
		xor ^= v
	}
	for i := 1; i <= len(nums); i++ {
		xor ^= i
	}

	lowBit := xor & -xor
	num1, num2 := 0, 0
	for _, v := range nums {
		if v&lowBit != 0 {
			num1 ^= v
		} else {
			num2 ^= v
		}
	}
	for i := 1; i <= len(nums); i++ {
		if i&lowBit != 0 {
			num1 ^= i
		} else {
			num2 ^= i
		}
	}

	for _, v := range nums {
		if v == num1 {
			return []int{num1, num2}
		}
	}

	return []int{num2, num1}
}
